### DISCUSSION TITLE
Weight decay for token embeddings

### DISCUSSION THREAD

**Discussion by MasterSkepticista** [karpathy/build-nanogpt#68] [Q&A]
Hi,

Why did we not exclude `embedding` weights from weight decay?
https://github.com/karpathy/build-nanogpt/blob/6104ab1b53920f6e2159749676073ff7d815c1fa/train_gpt2.py#L185-L186

Re: https://github.com/karpathy/minGPT/pull/24#issuecomment-679316025

This may also explain divergence in longer runs (?)

**Comment by shayanjl** [comment]
A likely reason is just that the script uses a very simple rule: apply weight decay to all parameters with `dim() >= 2`, which includes embedding matrices as well as linear weights. That is a convenient heuristic, not a statement that embeddings must be decayed.

In many transformer implementations, token embeddings are often excluded from weight decay, especially together with biases and normalization parameters, because they behave more like lookup tables than ordinary weight matrices. So your concern is reasonable.

Whether this causes divergence depends on the setup, but it can definitely change training behavior over long runs. So the short answer is: they were probably not intentionally singled out, they were just caught by the generic `dim() >= 2` rule. If you want the more standard choice, excluding embedding weights from decay is a sensible modification.

**Comment by SumanthSetty7777** [comment]
@MasterSkepticista One more reason can be that we are sharing the same weights with lm_head.
[ref](https://github.com/karpathy/build-nanogpt/blob/6104ab1b53920f6e2159749676073ff7d815c1fa/train_gpt2.py#L94)
        `self.transformer.wte.weight = self.lm_head.weight`
